File: src/main/java/de/bst/five_cities/opengeodb/GeoObject.java
package de.bst.five_cities.opengeodb;

import java.io.PrintStream;
import java.io.Serializable;
import java.util.Arrays;
import java.util.Map;
import java.util.Map.Entry;
import java.util.NoSuchElementException;

import org.apache.spark.broadcast.Broadcast;

public class GeoObject implements Serializable {

	private static final long serialVersionUID = -6810268884235023962L;

	private int id;

	private String[] splits;

	private Map<String, Integer> mapping;

	public GeoObject(String line, Map<String, Integer> mapping) {
		splits = line.split("\t");
		this.mapping = mapping;
		this.id = Integer.parseInt(splits[0]);
	}

	public GeoObject(String line, Broadcast<Map<String, Integer>> mapping) {
		this(line, mapping.value());
	}

	public GeoDistance distanceTo(GeoObject otherGeoObject) {
		return new GeoDistance(this, otherGeoObject);
	}

	public double dist(GeoObject other) {
		return GeoDistance.dist(this, other);
	}

	public void writeGeoEntry(PrintStream out) {
		for (Entry<String, Integer> key2index : mapping.entrySet()) {
			Integer index = key2index.getValue();
			if (index < splits.length)
				out.println("\t" + key2index.getKey() + ": " + splits[index]);
		}
		out.println();
	}

	public int getId() {
		return id;
	}

	public String getName() {
		return getString("name");
	}

	public String getString(int index) {
		return index < splits.length ? splits[index] : null;
	}

	public String getString(String key) {
		return getString(index(key));
	}

	private int index(String key) {
		Integer index = mapping.get(key);
		if (index != null)
			return index;
		else
			throw new NoSuchElementException("No mapping for " + key);
	}

	public double getEinwohner() {
		String key = "einwohner";
		return parseDouble(key);
	}

	public double getFlaeche() {
		String key = "flaeche";
		return parseDouble(key);
	}

	public boolean hasPosition() {
		return getString("lat") != null && getString("lon") != null //
				&& !getString("lat").isEmpty() && !getString("lon").isEmpty() //
				&& Double.isFinite(getLatitude()) && Double.isFinite(getLongtitude());
	}

	public double getLatitude() {
		return Double.parseDouble(getString("lat"));
	}

	public double getLongtitude() {
		return Double.parseDouble(getString("lon"));
	}

	public int getLevel() {
		return parseInt("level");
	}

	private long parseLong(String key) {
		String asString = getString(key);
		return asString.isEmpty() ? 0 : Long.parseLong(asString);
	}

	private double parseDouble(String key) {
		String asString = getString(key);
		return asString.isEmpty() ? 0.0 : Double.parseDouble(asString);
	}

	private int parseInt(String key) {
		String asString = getString(key);
		return asString == null || asString.isEmpty() ? 0 : Integer.parseInt(asString);
	}

	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder("GeoObject [" + getId() + "/" + getString("name"));
		Arrays.asList("einwohner", "lat", "lon", "level", "typ").stream().forEach(key -> {
			sb.append(" ");
			sb.append(key);
			sb.append("=");
			sb.append(getString(key));
		});
		sb.append("]");
		return sb.toString();
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + id;
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		GeoObject other = (GeoObject) obj;
		if (id != other.id)
			return false;
		return true;
	}

}

File: src/main/scala/de/bst/five_cities/opengeodb/GeoObject.scala
package de.bst.five_cities.opengeodb

import java.io.PrintStream

import org.apache.spark.broadcast.Broadcast

@SerialVersionUID(-6810268884235023962L)
class GeoObject(val id: Int, val splits: Seq[String], val mapping: Map[String, Int]) extends Serializable {
  val PRIME = 31

  def distanceTo(otherGeoObject: GeoObject): GeoDistance = GeoDistance(this, otherGeoObject)

  def dist(other: GeoObject): Double = GeoDistance.dist(this, other)

  def writeGeoEntry(out: PrintStream) = mapping foreach {
    entry =>
      val index: Integer = entry._2
      if (entry._2 < splits.length) out.println(s"\t${entry._1}: ${splits(entry._2)}")
  }

  def getName: String = getString("name")

  def getEinwohner: Double = parseDouble("einwohner")

  def getFlaeche: Double = parseDouble("flaeche")

  def hasPosition: Boolean =
    !getString("lat").isEmpty && !getString("lon").isEmpty && !getLatitude.isInfinite && !getLongtitude.isInfinite

  def getLatitude: Double = getString("lat").toDouble

  def getLongtitude: Double = getString("lon").toDouble

  def getLevel: Int = parseInt("level")

  override def toString: String = {
    val children = List("einwohner", "lat", "lon", "level", "typ") map (f => s"$f=${getString(f)}") mkString " "
    s"GeoObject [$id/${getString("name")} $children]"
  }

  override def hashCode: Int = PRIME + id

  private def getString(index: Int): String = if (index < splits.length) splits(index) else ""

  private def getString(key: String): String = getString(index(key))

  private def index(key: String): Int = {
    val index = mapping get key
    if (index.isDefined) index.get
    else throw new NoSuchElementException(s"No mapping for $key")
  }

  private def parseLong(key: String) = {
    val asString = getString(key)
    if (asString.isEmpty) 0l
    else asString.toLong
  }

  private def parseDouble(key: String) = {
    val asString = getString(key)
    if (asString.isEmpty) .0
    else asString.toDouble
  }

  private def parseInt(key: String) = {
    val asString = getString(key)
    if (asString.isEmpty) 0
    else asString.toInt
  }
}

object GeoObject {
  def apply(line: String, mapping: Map[String, Int]): GeoObject = {
    val splits = line split "\t"
    new GeoObject(splits(0).toInt, splits, mapping)
  }

  def apply(line: String, mapping: Broadcast[Map[String, Int]]): GeoObject = apply(line, mapping.value)
}
